File: 422/valid-word-square.py
# 這題超簡單：對角對稱的矩陣。如果字不夠長，就不要看它。
class Solution:
    def validWordSquare(self, words: List[str]) -> bool:
        N = len(words)
        for i in range(N):
            N2 = len(words[i]) # 因每個字的長度可能不同
            for j in range(N2): # words[i][j]一定有
                # 但 words[j][i] 要測長度
                # 列出不合格的條件:列不夠 or 不夠長 or 不相等
                if N<=j or len(words[j])<=i or words[i][j] != words[j][i]:
                    return False
        return True
# case 15/35: ["abc","b"]

File: 422/valid-word-square.cpp
// LeetCode 422. Valid Word Square
// 檢查 words 裡 words[i][j]與words[j][i]全部相同
// 不一定是「方塊」哦！
// case 33/35: ["abc", "b", "c"] 竟然有「奇怪形狀」的 words
class Solution {
public:
    bool validWordSquare(vector<string>& words) {
        for (int i=0; i<words.size(); i++) {
            for (int j=0; j<words[i].size(); j++) { 
                // words[j][i] 跑到外面，失敗
                if (j>=words.size() || i>=words[j].size()) return false;
                // 只要有任何不同，就失敗
                if (words[i][j] != words[j][i]) return false;
            }
        }
        // 都沒有失敗，就成功
        return true;
    }
};
